geomesa-lambda/geomesa-lambda-datastore/src/main/scala/org/locationtech/geomesa/lambda/stream/kafka/KafkaFeatureCache.scala

```diff
@@ -0,0 +1,260 @@
+/***********************************************************************
+ * Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.lambda.stream.kafka
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
+import java.util.concurrent.locks.ReentrantLock
+
+import com.typesafe.scalalogging.LazyLogging
+import org.joda.time.{DateTime, DateTimeZone}
+import org.locationtech.geomesa.lambda.stream.OffsetManager.OffsetListener
+import org.locationtech.geomesa.lambda.stream.kafka.KafkaFeatureCache.{ExpiringFeatureCache, ReadableFeatureCache, WritableFeatureCache}
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * Locally cached features
+  */
+class KafkaFeatureCache(topic: String) extends WritableFeatureCache with ReadableFeatureCache
+    with ExpiringFeatureCache with OffsetListener with LazyLogging {
+
+  // map of feature id -> current feature
+  private val features = new ConcurrentHashMap[String, SimpleFeature]
+
+  // technically we should synchronize all access to the following arrays, since we expand them if needed;
+  // however, in normal use it will be created up front and then only read.
+  // if partitions are added at runtime, we may have synchronization issues...
+
+  // array, indexed by partition, of queues of (offset, create time, feature), sorted by offset
+  private val queues = ArrayBuffer.empty[(ReentrantLock, java.util.ArrayDeque[(Long, Long, SimpleFeature)])]
+  private val offsets = ArrayBuffer.empty[AtomicLong]
+
+  private val debug = logger.underlying.isDebugEnabled()
+
+  override def partitionAssigned(partition: Int, offset: Long): Unit = {
+    logger.debug(s"Partition assigned: [$topic:$partition:$offset]")
+    ensurePartition(partition, offset)
+  }
+
+  override def get(id: String): SimpleFeature = features.get(id)
+
+  override def all(): Iterator[SimpleFeature] = {
+    import scala.collection.JavaConverters._
+    features.values.iterator.asScala
+  }
+
+  override def add(feature: SimpleFeature, partition: Int, offset: Long, created: Long): Unit = {
+    if (offsets(partition).get < offset) {
```

Comment:
we'd still want to put the feature in the cache if the ttl has passed, as otherwise it won't be picked up by the persistence run and written to accumulo